framework/decode/vulkan_replay_consumer_base.cpp

```diff
@@ -432,108 +429,93 @@ void VulkanReplayConsumerBase::ProcessCreateHardwareBufferCommand(
         ahb_info.memory_id           = memory_id;
         ahb_info.hardware_buffer     = buffer;
 
-        void* data = nullptr;
-        result     = -1;
+        result = -1;
 
         std::vector<format::HardwareBufferPlaneInfo> replay_plane_info;
 
         // The multi-plane functions are declared for API 26, but are only available to link with API 29.  So, this
         // could be turned into a run-time check dependent on dlsym returning a valid pointer for
         // AHardwareBuffer_lockPlanes.
 #if __ANDROID_API__ >= 29
-        AHardwareBuffer_Planes ahb_planes;
-        result = AHardwareBuffer_lockPlanes(buffer, AHARDWAREBUFFER_USAGE_CPU_WRITE_OFTEN, -1, nullptr, &ahb_planes);
-        if (result == 0)
+        if (desc.usage & AHARDWAREBUFFER_USAGE_CPU_WRITE_MASK)
         {
-            data = ahb_planes.planes[0].data;
+            AHardwareBuffer_Planes ahb_planes;
+            result =
+                AHardwareBuffer_lockPlanes(buffer, AHARDWAREBUFFER_USAGE_CPU_WRITE_OFTEN, -1, nullptr, &ahb_planes);
+            if (result == 0)
+            {
+                void* data = ahb_planes.planes[0].data;
+
+                for (uint32_t i = 0; i < ahb_planes.planeCount; ++i)
+                {
+                    format::HardwareBufferPlaneInfo ahb_plane_info;
+                    ahb_plane_info.offset =
+                        reinterpret_cast<uint8_t*>(ahb_planes.planes[i].data) - reinterpret_cast<uint8_t*>(data);
+                    ahb_plane_info.pixel_stride = ahb_planes.planes[i].pixelStride;
+                    ahb_plane_info.row_pitch    = ahb_planes.planes[i].rowStride;
+                    replay_plane_info.emplace_back(std::move(ahb_plane_info));
+                }
 
-            for (uint32_t i = 0; i < ahb_planes.planeCount; ++i)
+                if (AHardwareBuffer_unlock(buffer, nullptr) != 0)
+                {
+                    GFXRECON_LOG_ERROR("AHardwareBuffer_unlock failed for AHardwareBuffer object (Buffer ID = %" PRIu64
+                                       ", Memory ID = %" PRIu64 ")",
+                                       buffer_id,
+                                       memory_id);
+                }
+            }
+            else
             {
-                format::HardwareBufferPlaneInfo ahb_plane_info;
-                ahb_plane_info.offset =
-                    reinterpret_cast<uint8_t*>(ahb_planes.planes[i].data) - reinterpret_cast<uint8_t*>(data);
-                ahb_plane_info.pixel_stride = ahb_planes.planes[i].pixelStride;
-                ahb_plane_info.row_pitch    = ahb_planes.planes[i].rowStride;
-                replay_plane_info.emplace_back(std::move(ahb_plane_info));
+                GFXRECON_LOG_WARNING("AHardwareBuffer_lockPlanes failed: AHardwareBuffer_lock will be used instead");
```

Comment:
Is this warning obsolete now?

Reply:
No it should still be valid.
In the `__ANDROID_API__ >= 29` case I left the AHB lock as is. In this case it actually kinda accesses the buffer's data (`ahb_planes.planes[i].data`) although just to calculate offsets.
I wonder if there's a better way to calculate those offsets without locking but I am a bit reluctant to tamper with the `__ANDROID_API__ >= 29` code as it deals with multi-planar images and I don't know if we have a test case to test potential changes.

Reply:
It looks like you deleted the call to AHardwareBuffer_lock.  Am I misreading the diff?

Reply:
I don't see an instance of `_lock` in ProcessCreateHardwareBufferCommand in framework/decode/vulkan_replay_consumer_base.cpp in your branch.

Reply:
Oops, sorry, you're right.